### ISSUE TITLE
Speed of model generation

### ISSUE THREAD

**Issue by LqNoob** [serkansulun/midi-emotion#1]
Hi, @serkansulun. Thank you for your great work.
The quality of the music generated by the model is really very good. 
However, when the batch size is 1(without considering the effect of batch size), the speed of model generation is close to 1:1, which seems to be a bit slow compared to other music generation models. Is this problem mainly related to the representation of music, or is it caused by the nature of autoregression itself? 
Will you improve the model speed in the future?

**Comment by serkansulun** [comment]
Hi, and thank you. 

It is mainly related to model size and input size. In my preliminary experiments, I have seen better results whenever I increased the model size or input size, and no overfitting. So in order to obtain the best quality output, I used the largest models and longest input that I can fit into the GPU. So if you want to quickly increase the generation speed, you can reduce the value of `--max_input_len` in `generate.py`. Alternatively, you can train a smaller model, by reducing the value of `--n_layer` in `train.py`.

**Comment by LqNoob** [comment]
Thank you for your quick reply!
1. When I read the code carefully, I found that the model training part did not use fine-tune form, but the paper proposed to use pre-training weights for unconditional generative model (vanilla model). So I would like to ask, do I need unconditional training weights when training a conditional generation model.
2. Also, when you train the continuous_concat model, how much does the loss drop or the step reach to indicate that the model has converged to a better result?

**Comment by serkansulun** [comment]
> Thank you for your quick reply!
> 
>     1. When I read the code carefully, I found that the model training part did not use fine-tune form, but the paper proposed to use pre-training weights for unconditional generative model (vanilla model). So I would like to ask, do I need unconditional training weights when training a conditional generation model.

You can train conditional models from scratch, you don't need a pretrained model. But in case you want to do transfer learning, I added a script for that in `src/models/transfer_model.py`

>     2. Also, when you train the continuous_concat model, how much does the loss drop or the step reach to indicate that the model has converged to a better result?

I don't have the previous outputs, but you can see Table 2 in the paper. Negative log-likelihood is the same thing as loss, so continuous_concat model has 0.037 lower evaluation loss compared to the vanilla model.

**Comment by LqNoob** [comment]
Thank you very much for your answer, it solved some of my doubts in the training process.
But I have one more question about the Russell's circumplex model of affect consisting of valence and arousal dimensions. Looking at the range of (valence, arousal) in the dataset in two-dimensional coordinates, the final (valence, arousal) are in the form of squares, which do not seem to correspond to the Russell's circumplex model. 
So I can't understand how the final emotional labels (happy, sad, angry and calm) were obtained and what they used as a reference.

**Comment by serkansulun** [comment]
What do you mean by squares? It also doesn't use categorical labels like happy, sad, angry and calm. Only the discrete_token model uses discrete conditions such as A2 or V-1, which correspond to bins from the valence and arousal axes. The other conditional models use values between -1 and 1.

**Comment by LqNoob** [comment]
> What do you mean by squares? 

The following figure is from [this article](http://stephane-gobron.net/Core/Publications/Papers/2010_Engage2010_asymFacialExpres.pdf). From the valence and arousal axes that both take values in the range [-1,1], it can be found that the (valence, arousal) points form a square in plane coordinates. It can be easily found from the diagonal values that there is no correspondence with Russell's circumplex model of affect. 


<img width="829" alt="Emotional model - 3" src="https://user-images.githubusercontent.com/28683383/183324788-fce25315-5315-40e6-8937-9035a33a0f7c.png">

**Comment by serkansulun** [comment]
Thanks, you're right, they're slightly different. Still, it doesn't change much in practice.

**Comment by serkansulun** [comment]
What do you mean by squares? It also doesn't use categorical labels like happy, sad, angry and calm. Only the discrete_token model uses discrete conditions such as A2 or V-1, which correspond to bins from the valence and arousal axes. The other conditional models use values between -1 and 1.

________________________________
